tests/cupy_tests/linalg_tests/test_decomposition.py

```diff
@@ -177,7 +173,15 @@ def check_usv(self, shape, dtype):
             numpy.testing.assert_allclose(vj_gpu, v_sign * vj_cpu, atol=1e-4)
             numpy.testing.assert_allclose(abs(u_sign), 1, atol=1e-4)
             numpy.testing.assert_allclose(abs(v_sign), 1, atol=1e-4)
-            numpy.testing.assert_allclose(u_sign, v_sign, atol=1e-4)
+            # we don't compare u_sign with v_sign directly, as the sign
+            # convention is implementation dependent
```

Comment:
@toslunar It seems you were the last person touching this test (https://github.com/cupy/cupy/pull/2338). Could you comment on this change? On HIP the sign convention seems different from CUDA (differs by -1 in some cases), but the reconstructed matrix is identical (see the added test below). I think this is allowed.

Reply:
Could you give me an example input/output of `svd` by HIP?  I couldn't figure out whether my error was in math or in coding.  The sign convention is even different between CUDA and numpy.

If the reconstruction test is added, it seems nice to change the for-block to
```python
testing.assert_allclose(abs(u_gpu), abs(u_cpu), atol=1e-4)
testing.assert_allclose(abs(vh_gpu), abs(vh_cpu), atol=1e-4)
```
because there are unitarity assertions.

Reply:
Hi @toslunar I looked into it, and there's only one problematic test, with the matrix being `[[0 1], [0 1]]` and of int type. But I think this is a singular matrix leading to zero singular value, so the corresponding eigenvector's sign can be undetermined. Should we simply do this instead of removing the check completely?
```python
            try:
                numpy.testing.assert_allclose(u_sign, v_sign, atol=1e-4)
            except AssertionError:
                if s_cpu[j] != 0.:
                    raise
```

Reply:
> ```
> FAILED tests/cupy_tests/linalg_tests/test_product.py::TestDot_param_0_{shape=((2, 3, 4), (3, 4, 2)), trans_a=True, trans_b=True}::test_dot - AssertionError: 
> FAILED tests/cupy_tests/linalg_tests/test_product.py::TestDot_param_1_{shape=((2, 3, 4), (3, 4, 2)), trans_a=True, trans_b=False}::test_dot - AssertionError: 
> FAILED tests/cupy_tests/linalg_tests/test_product.py::TestDot_param_2_{shape=((2, 3, 4), (3, 4, 2)), trans_a=False, trans_b=True}::test_dot - AssertionError: 
> FAILED tests/cupy_tests/linalg_tests/test_product.py::TestDot_param_3_{shape=((2, 3, 4), (3, 4, 2)), trans_a=False, trans_b=False}::test_dot - AssertionError: 
> FAILED tests/cupy_tests/linalg_tests/test_product.py::TestProduct::test_multidim_inner - AssertionError: 
> FAILED tests/cupy_tests/linalg_tests/test_product.py::TestProduct::test_tensordot - AssertionError: 
> FAILED tests/cupy_tests/linalg_tests/test_product.py::TestProduct::test_transposed_dot - AssertionError: 
> FAILED tests/cupy_tests/linalg_tests/test_product.py::TestProduct::test_transposed_dot_with_out - AssertionError: 
> FAILED tests/cupy_tests/linalg_tests/test_product.py::TestProduct::test_transposed_higher_order_inner - AssertionError: 
> FAILED tests/cupy_tests/linalg_tests/test_product.py::TestProduct::test_transposed_tensordot - AssertionError: 
> ```
> 
> I will create an issue to track the `dot` failures, as it seems to be related to the specialized integer kernel.

This will be fixed in #4512!

Reply:
Thanks a lot, @leofang.

Let me summarize the uniqueness of SVD "up to rotation".

- For each nonzero singular value, a rotation is applied to the corresponding pair of vectors in `u` and `v`.  A 1-dim rotation is a sign.  If the singular value is repeated, the rotation is not necessarily signs.
- For the remaining part (if a singular value is 0 or `full_matrices=True`), a rotation applies to the corresponding vectors in `u` and another rotation applies to those in `v`.

Maybe I assumed in #2338 that the singular values are nonzero and distinct by `fix_random`, but there is a case that a singular value is 0.  (In my previous post, `k = rank(A)` versus `k = min(A.shape)` confused me.)

If we update the test by such skips, it should skip singular values that are **close to 0 or close to another**.  However, I'd like only to replace the loop with the reconstruction test (`a_gpu_usv`), as long as the tolerances are kept loose (`1e-4`).

Reply:
Ah, good point, thanks @toslunar I haven't thought about the issue of degeneracy. OK I also think it's best to to just ensure the matrix can be reconstructed, which is really the only thing SVD could guarantee regardless of implementation detail. Will clean it up later.

Reply:
Done in 232454a.